library/core/src/num/mod.rs

```diff
@@ -688,7 +688,8 @@ impl u8 {
     #[rustc_const_stable(feature = "const_ascii_ctype_on_intrinsics", since = "1.47.0")]
     #[inline]
     pub const fn is_ascii_hexdigit(&self) -> bool {
-        matches!(*self, b'0'..=b'9' | b'A'..=b'F' | b'a'..=b'f')
+        // Bitwise or converts A-Z to a-z, avoiding need for branches in compiled code.
```

Comment:
Minor nit, but it feels like the lack of branches still isn't clear to the passive viewer since `'0'..='9'` and `'a'..='f'` are still two separate ranges with a gap in the middle.

Oh, and uh, that `||` is looking suspiciously not-branchless. Since it ends up being so in the resulting assembly, perhaps it could be reworded as `|`?

Reading the compiled output:
```asm
example::char_is_hex_2:
        mov     eax, dword ptr [rdi]
        lea     ecx, [rax - 48]
        cmp     ecx, 10
        setb    cl
        or      eax, 32
        add     eax, -97
        cmp     eax, 6
        setb    al
        or      al, cl
        ret
```

Which is effectively:
```rust
(*self - 48 < 10) | ((*self | 0x20) - 97 < 6)
```

It's less that there are "fewer" branches in this code, but more that going from three ranges to two triggers a threshold in LLVM's side that makes it decide that branches are no longer worth it, and it removes them.

So... tying this all together, maybe the real point is not that it's branchless by itself, but that there are fewer computations overall _and_ LLVM is likely to optimise it without branches as a result.

Reply:
That seems reasonable, I can do one where all the optimization is done by hand to make it clear what the resulting assembly should be.

Reply:
Sorry for pushing churn on you here, but I'm actually going to ask that you aim for a middle ground where the code is still as obvious as possible while still producing branchless assembly.

Particularly, there's no need to replace the range changes with wrapping_sub manually, as LLVM will do that quite reliably: <https://rust.godbolt.org/z/rPaxbf1o7>

It's normal in the library for the code to *not* be in the form of the expected assembly.  For examples, `is_power_of_two` is phrased using `count_ones`, not the well-known bitwise tricks

https://github.com/rust-lang/rust/blob/1ca6777c014813e3bdb98d155562fc3d111d86dd/library/core/src/num/uint_macros.rs#L2132-L2134

as that leaves it up to LLVM to generate the best assembly sequence -- which *will* be the `(x != 0) && (x & (x-1)) == 0` on i386, but libcore doesn't need to know that.

So please go back to the version with the `const`s you had in a previous iteration.  And consider making a local for the lower-cased version of the character -- that would help localize the comment.  Maybe something like

```rust
        // Bitwise or converts A-Z to a-z, allowing checking for the letters with a single range.
        let lower = *self as u32 | num::ASCII_CASE_MASK;
        const A: u32 = 'a' as u32;
        const F: u32 = 'f' as u32;
        // Not using logical or because the branch isn't worth it
        self.is_ascii_digit() | matches!(*self as u32 | 0x20, A..=F)
```

You could also consider splitting the hex_letter part into a separate (non-`pub`) function so that `is_ascii_hexdigit` becomes just
```rust
        // Not using logical or because the branch isn't worth it
        self.is_ascii_digit() | self.is_ascii_hexletter()
```
to isolate the conversion-and-range stuff to its own thing.

@rustbot author

Reply:
Hmm, it's just something about the *or pattern* in `matches!`.  This is also branchless: <https://rust.godbolt.org/z/9xqT38rz9>
```rust
fn is_ascii_digit(c: u8) -> bool {
    matches!(c, b'0'..=b'9')
}

fn is_ascii_hexletter(c: u8) -> bool {
    matches!(c, b'A'..=b'F') | matches!(c, b'a'..=b'f')
}

pub fn is_ascii_hexdigit(c: u8) -> bool {
    is_ascii_digit(c) | is_ascii_hexletter(c)
}
```

(And yup, it works for `char` too: <https://rust.godbolt.org/z/WvfsjMG6a>.)

Reply:
Loving the investigative work here. Perhaps this is an actual codegen issue, since `matches!(x, P) | matches!(x, Q)` should ideally generate code rather close to `matches!(x, P | Q)`. It feels like more of a codegen quirk than an LLVM optimizer quirk, but I could be wrong.

Reply:
I dug in some more, and it's actually more interesting than I expected!

It looks like with the or-pattern it's tripping a different technique in LLVM. <https://rust.godbolt.org/z/zdW1T43T3>

Starting with the nice obvious
```rust
pub fn is_ascii_hexletter(c: char) -> bool {
    matches!(c, 'A'..='F' | 'a'..='f')
}
```
SimplifyCfg and SROA get it down to the nice simple short-circuiting
```llvm
define noundef zeroext i1 @example::is_ascii_hexletter(i32 noundef %c) unnamed_addr {
start:
  %_4 = icmp ule i32 65, %c
  %_5 = icmp ule i32 %c, 70
  %or.cond = and i1 %_4, %_5
  br i1 %or.cond, label %bb5, label %bb2

bb2:                                              ; preds = %start
  %_2 = icmp ule i32 97, %c
  %_3 = icmp ule i32 %c, 102
  %or.cond1 = and i1 %_2, %_3
  br i1 %or.cond1, label %bb5, label %bb4

bb5:                                              ; preds = %bb2, %start
  br label %bb6

bb4:                                              ; preds = %bb2
  br label %bb6

bb6:                                              ; preds = %bb4, %bb5
  %.0 = phi i8 [ 1, %bb5 ], [ 0, %bb4 ]
  %0 = trunc i8 %.0 to i1
  ret i1 %0
}
```
and InstCombine does the "I know how to simplify range checks like that" rewrite:
```llvm
define noundef zeroext i1 @example::is_ascii_hexletter(i32 noundef %c) unnamed_addr {
start:
  %0 = add i32 %c, -65
  %1 = icmp ult i32 %0, 6
  br i1 %1, label %bb5, label %bb2

bb2:                                              ; preds = %start
  %2 = add i32 %c, -97
  %3 = icmp ult i32 %2, 6
  br i1 %3, label %bb5, label %bb4

bb5:                                              ; preds = %bb2, %start
  br label %bb6

bb4:                                              ; preds = %bb2
  br label %bb6

bb6:                                              ; preds = %bb4, %bb5
  %.0 = phi i1 [ true, %bb5 ], [ false, %bb4 ]
  ret i1 %.0
}
```
But then something really interesting happens, and `SimplifyCfg` says "wait, that looks like a lookup table!", giving
```llvm
define noundef zeroext i1 @example::is_ascii_hexletter(i32 noundef %c) unnamed_addr {
start:
  switch i32 %c, label %bb4 [
    i32 102, label %bb6
    i32 101, label %bb6
    i32 100, label %bb6
    i32 99, label %bb6
    i32 98, label %bb6
    i32 97, label %bb6
    i32 70, label %bb6
    i32 69, label %bb6
    i32 68, label %bb6
    i32 67, label %bb6
    i32 66, label %bb6
    i32 65, label %bb6
  ]

bb4:                                              ; preds = %start
  br label %bb6

bb6:                                              ; preds = %start, %start, %start, %start, %start, %start, %start, %start, %start, %start, %start, %start, %bb4
  %.0 = phi i1 [ false, %bb4 ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ], [ true, %start ]
  ret i1 %.0
}
```
Then later SimplifyCfg looks at that *again* and say "wait, that's a weird switch, how about I do that with a shift instead?" by encoding the table into an `i38` (because `'A'` through `'f'` [is 38 values](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f7d86be1f8b8c5d1b42392d54d3e7d96)):
```llvm
define noundef zeroext i1 @example::is_ascii_hexletter(i32 noundef %c) unnamed_addr {
start:
  %switch.tableidx = sub i32 %c, 65
  %0 = icmp ult i32 %switch.tableidx, 38
  %switch.cast = zext i32 %switch.tableidx to i38
  %switch.shiftamt = mul i38 %switch.cast, 1
  %switch.downshift = lshr i38 -4294967233, %switch.shiftamt
  %switch.masked = trunc i38 %switch.downshift to i1
  %.0 = select i1 %0, i1 %switch.masked, i1 false
  ret i1 %.0
}
```
which is *also* branchless.  Just not quite as good a way.